File: Deliverables/HEAVYROUTE/src/main/java/com/heavyroute/resources/dto/VehicleDTO.java
package com.heavyroute.resources.dto;

import com.heavyroute.resources.enums.VehicleStatus;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Data Transfer Object (DTO) per la rappresentazione dei dati tecnici e operativi di un veicolo.
 * <p>
 * Include le validazioni necessarie per garantire l'integrità dei dati della flotta,
 * prevenendo l'inserimento di mezzi con parametri fisici non validi (es. pesi o dimensioni negative)
 *.
 * </p>
 * * @author Heavy Route Team
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
public class VehicleDTO {

    /**
     * Targa del veicolo.
     * Utilizzata come riferimento principale nelle comunicazioni operative e nei documenti di viaggio.
     * Non può essere vuota o composta solo da spazi.
     */
    @NotBlank(message = "La targa del veicolo è obbligatoria")
    private String licensePlate;

    /**
     * Marca e modello del veicolo (es. "Iveco Stralis 500").
     * Utilizzato per l'identificazione rapida del mezzo da parte del personale.
     */
    @NotBlank(message = "Il modello del veicolo è obbligatorio")
    private String model;

    /**
     * Portata massima utile del mezzo espressa in chilogrammi.
     * Questo valore è fondamentale per il controllo di compatibilità:
     * il peso del carico deve essere inferiore o uguale a questo parametro.
     */
    @NotNull(message = "La portata massima è obbligatoria")
    @Positive(message = "La portata massima deve essere un valore positivo")
    private Double maxLoadCapacity;

    /**
     * Altezza massima del veicolo in metri (incluso il pianale di carico).
     * Parametro critico per la verifica della percorribilità sotto tunnel o ponti.
     */
    @NotNull(message = "L'altezza massima è obbligatoria")
    @Positive(message = "L'altezza massima deve essere un valore positivo")
    private Double maxHeight;

    /**
     * Larghezza massima del veicolo in metri.
     * Determina la necessità di scorte tecniche se supera i limiti stradali standard.
     */
    @NotNull(message = "La larghezza massima è obbligatoria")
    @Positive(message = "La larghezza massima deve essere un valore positivo")
    private Double maxWidth;

    /**
     * Lunghezza massima del veicolo in metri.
     * Parametro utilizzato per verificare la manovrabilità nelle curve strette.
     */
    @NotNull(message = "La lunghezza massima è obbligatoria")
    @Positive(message = "La lunghezza massima deve essere un valore positivo")
    private Double maxLength;

    /**
     * Stato operativo attuale del veicolo (es. AVAILABLE, BUSY, MAINTENANCE).
     * Determina se il mezzo può essere assegnato a un nuovo viaggio.
     */
    private VehicleStatus status;
}

File: Deliverables/heavyroute_app/lib/features/planner/presentation/model/vehicle_dto.dart
/**
 * Data Transfer Object (DTO) per la rappresentazione dei dati tecnici e operativi di un veicolo.
 * <p>
 * Mappa i parametri fisici del mezzo (pesi e dimensioni) utilizzati dal sistema
 * per verificare la compatibilità con le infrastrutture stradali.
 * </p>
 * @author Roman
 */
class VehicleDTO {
  /** Targa univoca del veicolo. */
  final String licensePlate;

  /** Marca e modello del mezzo (es. "Iveco Stralis 500"). */
  final String model;

  /** Portata massima utile espressa in chilogrammi. */
  final double maxLoadCapacity;

  /** Altezza massima del veicolo in metri. */
  final double maxHeight;

  /** Larghezza massima del veicolo in metri. */
  final double maxWidth;

  /** Lunghezza massima del veicolo in metri. */
  final double maxLength;

  /** Stato operativo attuale (es. AVAILABLE, BUSY, MAINTENANCE). */
  final String? status;

  VehicleDTO({
    required this.licensePlate,
    required this.model,
    required this.maxLoadCapacity,
    required this.maxHeight,
    required this.maxWidth,
    required this.maxLength,
    this.status,
  });

  /**
   * Factory method per la creazione di un'istanza da JSON proveniente dal backend.
   * @param json Mappa contenente i dati del veicolo.
   * @return Un'istanza di {@link VehicleDTO}.
   */
  factory VehicleDTO.fromJson(Map<String, dynamic> json) {
    return VehicleDTO(
      licensePlate: json['licensePlate'] as String,
      model: json['model'] as String,
      maxLoadCapacity: (json['maxLoadCapacity'] as num).toDouble(),
      maxHeight: (json['maxHeight'] as num).toDouble(),
      maxWidth: (json['maxWidth'] as num).toDouble(),
      maxLength: (json['maxLength'] as num).toDouble(),
      status: json['status'] as String?,
    );
  }

  /**
   * Converte l'oggetto in formato JSON per l'invio alle API di gestione flotta.
   * @return Mappa serializzata.
   */
  Map<String, dynamic> toJson() => {
    'licensePlate': licensePlate,
    'model': model,
    'maxLoadCapacity': maxLoadCapacity,
    'maxHeight': maxHeight,
    'maxWidth': maxWidth,
    'maxLength': maxLength,
    'status': status,
  };
}